# Ejercicio de manipulación de datos

Para realizar este ejercicio vamos a utilizar el fichero `UCM-Acceso-2024-25-Anonimizado.csv` de la página https://www.universidata.es/datasets/ucm-acceso. Desde dicha ágina podemos acceder al **Diccionario de Datos** 
https://dimetrical.atlassian.net/wiki/spaces/UNC/pages/1608286213/Dataset+Acceso con la descripción de cada campo del fichero.



## Carga de datos y exploración del data frame


Vamos a descargamos los ficheros correspondientes al curso 2024-25 (UCM-Acceso-2024-25-Anonimizado.csv) de la web
https://www.universidata.es/datasets/ucm-acceso.
Lo dejamos en nuestra carpeta llamada `Data` y lo importamos como un data frame llamado `df`.



In [121]:
import pandas as pd

df = pd.read_csv("./Data/ucm-acceso-2024-25-anonimizado.csv")


/var/folders/gz/w9zmh03s61l021wmll3psm_c0000gn/T/ipykernel_10320/1471365357.py:3: DtypeWarning: Columns (35) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("./Data/ucm-acceso-2024-25-anonimizado.csv")


Al hacer la carga de los datos desde el fichero hasta el data frame, me ha salido una advertencia (***DtypeWarning***) de Pandas: una de las columnas del csv contiene datos de tipos diferentes (***Columns (35) have mixed types.***). Esto conviene explorarlo, empezamos mirando las dimensiones, es decir, el número de filas y columnas.

In [124]:
df.shape      # número de filas y columnas

(14019, 78)

Vemos el nombre de las columnas

In [127]:
df.columns

Index(['IMPORTANTE', 'cod_universidad', 'des_universidad', 'curso_academico',
       'cod_campus', 'des_campus', 'cod_centro', 'des_centro',
       'cod_naturaleza_centro', 'des_naturaleza_centro', 'cod_tipo_estudio',
       'des_tipo_estudio', 'cod_titulacion', 'des_titulacion',
       'cod_comunidad_centro_sec', 'des_comunidad_centro_sec',
       'cod_provincia_centro_sec', 'des_provincia_centro_sec',
       'cod_municipio_centro_sec', 'des_municipio_centro_sec',
       'lat_municipio_centro_sec', 'lon_municipio_centro_sec',
       'cod_naturaleza_centro_sec', 'des_naturaleza_centro_sec',
       'nota_admision', 'cod_pais_fin_estudio_acceso',
       'des_pais_fin_estudio_acceso', 'lat_pais_fin_estudio_acceso',
       'lon_pais_fin_estudio_acceso', 'cod_continente_fin_estudio_acceso',
       'des_continente_fin_estudio_acceso',
       'cod_agregacion_paises_fin_estudio_acceso',
       'des_agregacion_paises_fin_estudio_acceso', 'cod_estudio_acceso',
       'des_estudio_acceso', 'cod_e

Nos interesa la columna 35, para ello preguntamos por el index 35 de la lista que devuelve `df.columns`.

In [130]:
df.columns[35]   

'cod_especialidad_acceso'

Veo el contenido de la columna indexada por 35 y de nombre `cod_especialidad_acceso`.

In [133]:
df[['cod_especialidad_acceso']]

,cod_especialidad_acceso
0,NaN
1,0703
2,NaN
3,NaN
4,NaN
...,...
14014,NaN
14015,NaN
14016,204.0
14017,NaN


## Problema: Los códigos como 0703 no deberían tratarse como números.

***Si Pandas intenta interpretar el código 0703 como un número, 0703 se convierte en 703, perdiendo el 0 inicial.***


### Solución 1

Decir a Pandas que lea el archivo de otra manera para determinar los tipos de las columnas (`low_memory = False`).



In [136]:
df = pd.read_csv(
    "./Data/ucm-acceso-2024-25-anonimizado.csv",
    low_memory = False   
)

df[['cod_especialidad_acceso']]

,cod_especialidad_acceso
0,NaN
1,0703
2,NaN
3,NaN
4,NaN
...,...
14014,NaN
14015,NaN
14016,0204
14017,NaN


### Solución 2

Cambiar explícitamente el tipo de la columna que puede generar el conflicto (`dtype={"cod_especialidad_acceso": "category"}`) es la solución más adecuada, ya que nos permite controlar cómo interpreta Pandas los datos al leer el archivo.

En nuestro caso podemos cambiarla por `str` o `category`. Si es `str`entonces se tratan a los valores como cadenas de caracters y si es `category` entonces, se tratan a los valores como un conjunto limitado de categorías, aunque sea muy, muy, muy grande, es limitado.

In [139]:
df = pd.read_csv(
    "./Data/ucm-acceso-2024-25-anonimizado.csv",
    dtype={"cod_especialidad_acceso": "category"}
)

df[['cod_especialidad_acceso']]

,cod_especialidad_acceso
0,NaN
1,0703
2,NaN
3,NaN
4,NaN
...,...
14014,NaN
14015,NaN
14016,0204
14017,NaN


De hecho, creo que es conveniente mirar **todas** las columnas numéricas.

In [142]:
df.describe()

,cod_universidad,cod_campus,cod_centro,cod_naturaleza_centro,cod_titulacion,cod_comunidad_centro_sec,cod_provincia_centro_sec,cod_municipio_centro_sec,lat_municipio_centro_sec,lon_municipio_centro_sec,...,lat_pais_residencia,lon_pais_residencia,cod_continente_residencia,cod_agregacion_paises_residencia,cod_dedicacion,cod_familia_numerosa,cod_nivel_estudios_madre,cod_ocupacion_madre,cod_nivel_estudios_padre,cod_ocupacion_padre
count,14019.0,14019.000000,1.399300e+04,14019.000000,1.399300e+04,12265.000000,12265.000000,12265.000000,12265.000000,12265.000000,...,13955.000000,13955.000000,14018.000000,14018.000000,14019.000000,14019.000000,13310.000000,12567.000000,12925.000000,11925.000000
mean,10.0,1001.424281,2.802956e+07,1.082959,2.846460e+06,11.748879,27.773828,27858.815165,40.059775,-3.968224,...,40.119129,-3.846344,148.001783,44.640106,1.024752,1.249590,4.558753,5.363810,4.444874,4.840922
std,0.0,1.035894,5.425562e+03,0.352975,1.197797e+06,3.066593,6.900210,6910.073205,1.998705,2.079938,...,4.432447,11.043432,16.160801,41.000761,0.155374,0.497367,0.691911,3.967887,0.752792,3.842955
min,10.0,1001.000000,2.802673e+07,1.000000,2.500308e+06,1.000000,1.000000,1008.000000,27.883700,-17.911370,...,-40.900557,-106.346771,2.000000,15.000000,1.000000,1.000000,1.000000,0.000000,1.000000,0.000000
25%,10.0,1001.000000,2.802700e+07,1.000000,2.500692e+06,13.000000,28.000000,28068.000000,40.326420,-3.865741,...,40.463667,-3.749220,150.000000,39.000000,1.000000,1.000000,4.000000,2.000000,4.000000,2.000000
50%,10.0,1001.000000,2.802714e+07,1.000000,2.500829e+06,13.000000,28.000000,28079.000000,40.416690,-3.700346,...,40.463667,-3.749220,150.000000,39.000000,1.000000,1.000000,5.000000,4.000000,5.000000,4.000000
75%,10.0,1002.000000,2.802784e+07,1.000000,2.501382e+06,13.000000,28.000000,28113.000000,40.425940,-3.700346,...,40.463667,-3.749220,150.000000,39.000000,1.000000,1.000000,5.000000,9.000000,5.000000,8.000000
max,10.0,1012.000000,2.805392e+07,3.000000,7.001218e+06,19.000000,52.000000,52001.000000,43.686340,4.264315,...,61.924110,174.885971,150.000000,419.000000,2.000000,3.000000,5.000000,14.000000,5.000000,14.000000


Veo que ¿todas? las variable numéricas comienzan por `cod_` y deberían ser categóricas (NO str), si tengo dudas tengo que consultar la documentación:
https://dimetrical.atlassian.net/wiki/spaces/UNC/pages/1608286213/Dataset+Acceso

Cambio todos los tipos de las columnas que representan códigos a categóricas.

In [145]:
for columna in df.columns:
    if columna.startswith("cod_"):
        df[columna] = df[columna].astype("category")

df.describe()

,lat_municipio_centro_sec,lon_municipio_centro_sec,des_pais_fin_estudio_acceso,lat_pais_fin_estudio_acceso,lon_pais_fin_estudio_acceso,des_continente_fin_estudio_acceso,des_agregacion_paises_fin_estudio_acceso,anio_fin_estudio_acceso,anio_acceso_SUE,anio_nacimiento,lat_pais_nacionalidad,lon_pais_nacionalidad,lat_pais_residencia,lon_pais_residencia
count,12265.000000,12265.000000,0.0,0.0,0.0,0.0,0.0,1.0,14019.000000,14019.000000,13933.000000,13933.000000,13955.000000,13955.000000
mean,40.059775,-3.968224,NaN,NaN,NaN,NaN,NaN,2008.0,2023.768885,2004.735430,39.632385,-1.304952,40.119129,-3.846344
std,1.998705,2.079938,NaN,NaN,NaN,NaN,NaN,NaN,2.138472,4.242151,6.758627,22.420059,4.432447,11.043432
min,27.883700,-17.911370,NaN,NaN,NaN,NaN,NaN,2008.0,1973.000000,1943.000000,-40.900557,-106.346771,-40.900557,-106.346771
25%,40.326420,-3.865741,NaN,NaN,NaN,NaN,NaN,2008.0,2024.000000,2005.000000,40.463667,-3.749220,40.463667,-3.749220
50%,40.416690,-3.700346,NaN,NaN,NaN,NaN,NaN,2008.0,2024.000000,2006.000000,40.463667,-3.749220,40.463667,-3.749220
75%,40.425940,-3.700346,NaN,NaN,NaN,NaN,NaN,2008.0,2024.000000,2006.000000,40.463667,-3.749220,40.463667,-3.749220
max,43.686340,4.264315,NaN,NaN,NaN,NaN,NaN,2008.0,2024.000000,2008.000000,61.924110,174.885971,61.924110,174.885971


Pues no eran todas. Hay variables numéricas que denotan años, latitud y longitud. pero puedo quitar las descripciones `_des`.

In [153]:
for columna in df.columns:
    if columna.startswith("des_"):
        df[columna] = df[columna].astype("category")
        
df.describe()

,lat_municipio_centro_sec,lon_municipio_centro_sec,lat_pais_fin_estudio_acceso,lon_pais_fin_estudio_acceso,anio_fin_estudio_acceso,anio_acceso_SUE,anio_nacimiento,lat_pais_nacionalidad,lon_pais_nacionalidad,lat_pais_residencia,lon_pais_residencia
count,12265.000000,12265.000000,0.0,0.0,1.0,14019.000000,14019.000000,13933.000000,13933.000000,13955.000000,13955.000000
mean,40.059775,-3.968224,NaN,NaN,2008.0,2023.768885,2004.735430,39.632385,-1.304952,40.119129,-3.846344
std,1.998705,2.079938,NaN,NaN,NaN,2.138472,4.242151,6.758627,22.420059,4.432447,11.043432
min,27.883700,-17.911370,NaN,NaN,2008.0,1973.000000,1943.000000,-40.900557,-106.346771,-40.900557,-106.346771
25%,40.326420,-3.865741,NaN,NaN,2008.0,2024.000000,2005.000000,40.463667,-3.749220,40.463667,-3.749220
50%,40.416690,-3.700346,NaN,NaN,2008.0,2024.000000,2006.000000,40.463667,-3.749220,40.463667,-3.749220
75%,40.425940,-3.700346,NaN,NaN,2008.0,2024.000000,2006.000000,40.463667,-3.749220,40.463667,-3.749220
max,43.686340,4.264315,NaN,NaN,2008.0,2024.000000,2008.000000,61.924110,174.885971,61.924110,174.885971


¿Quedan más columnas de tipo numérico que necesitan ser categóricas?

Para detectar qué columnas que no son `category`, filtar las no `category` de `Dtype`.

In [155]:
df.loc[:, df.dtypes != "category"].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14019 entries, 0 to 14018
Data columns (total 17 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   IMPORTANTE                    14019 non-null  object 
 1   curso_academico               14019 non-null  object 
 2   lat_municipio_centro_sec      12265 non-null  float64
 3   lon_municipio_centro_sec      12265 non-null  float64
 4   nota_admision                 13365 non-null  object 
 5   lat_pais_fin_estudio_acceso   0 non-null      float64
 6   lon_pais_fin_estudio_acceso   0 non-null      float64
 7   anio_fin_estudio_acceso       1 non-null      float64
 8   ind_accede_al_SUE_por_1a_vez  14019 non-null  object 
 9   anio_acceso_SUE               14019 non-null  int64  
 10  ind_se_matricula_en_primero   14019 non-null  object 
 11  ind_curso_adaptacion          14019 non-null  object 
 12  anio_nacimiento               14019 non-null  int64  
 13  l

Veamos sus datos de uno en uno

In [157]:
df['IMPORTANTE']

0        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
1        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
2        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
3        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
4        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
                               ...                        
14014    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
14015    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
14016    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
14017    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
14018    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
Name: IMPORTANTE, Length: 14019, dtype: object

O todos a la vez 

In [164]:
df['ind_accede_al_SUE_por_1a_vez']

0        SÍ
1        SÍ
2        SÍ
3        SÍ
4        SÍ
         ..
14014    SÍ
14015    SÍ
14016    SÍ
14017    NO
14018    SÍ
Name: ind_accede_al_SUE_por_1a_vez, Length: 14019, dtype: object